simhash/__init__.py

```diff
@@ -64,36 +64,68 @@ def _tokenize(self, content):
         ans = self._slide(content)
         return ans
 
-    def build_by_text(self, content):
-        features = self._tokenize(content)
-        features = {k:sum(1 for _ in g) for k, g in groupby(sorted(features))}
-        return self.build_by_features(features)
-
-    def build_by_features(self, features):
+    def update(self, features):
         """
         `features` might be a list of unweighted tokens (a weight of 1
-                   will be assumed), a list of (token, weight) tuples or
-                   a token -> weight dict.
+                   will be assumed), a list of (token, weight) tuples, a
+                   token -> weight dict or a string.
         """
-        v = [0] * self.f
-        masks = [1 << i for i in range(self.f)]
+        if isinstance(features, (str, unicode)):
+            # If we receive a string, we will prepend our buffer, tokenize the
+            # result, and withhold the last token in our buffer.
+            if self.buffer:
+                features = self.buffer + features
+            if not features:
+                return
+            features = self._tokenize(features)
+            self.buffer = features.pop()
+            features = {k:sum(1 for _ in g) for k, g in groupby(sorted(features))}
         if isinstance(features, dict):
             features = features.items()
         for f in features:
             if isinstance(f, basestring):
-                h = self.hashfunc(f.encode('utf-8'))
+                h = f.encode('utf-8')
                 w = 1
             else:
                 assert isinstance(f, collections.Iterable)
-                h = self.hashfunc(f[0].encode('utf-8'))
+                h = f[0].encode('utf-8')
                 w = f[1]
+            self.features[h] += w
```

Comment:
If I understand correctly, `len(h)` is always 4 unless the user is doing their own feature extraction and passing in an iterable right? So, with deferred hashing (my optimization) we are optimized for the default case. If the user is advanced enough to do their own feature extraction, I would suggest to them that they hash their features before passing them in, and nullify the hashfunc. In other words, they can optimize their own non-default use case... Here is an example of a use-case that would require it's own optimization, and I think reflects your concern.

```python
# This will produce features with len() > 16...
features = open('/some/file').readlines()
sh = Simhash(features)
# So the user can optimize like so:
features = [int(md5(l).digest(), 16) for l in open('/some/file').readlines()]
sh = Simhash(features, hashfunc=lambda s: s)
```